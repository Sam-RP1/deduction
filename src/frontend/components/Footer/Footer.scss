@use '../../styles/base';
@import '../../styles/mixins/transition';
@import '../../styles/mixins/transform';

.footer {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    height: 80px;
    width: 100%;
    box-sizing: border-box;
    padding: 10px 16px;
    background-color: base.$secondary-black;
    @media screen and (min-width: 768px) {
        flex-direction: row-reverse;
        height: 50px;
    }

    &__info,
    &__settings {
        &--row {
            display: flex;
            flex-direction: row;
            justify-content: space-between;
            align-items: center;

            p {
                font-size: 1.1rem;
                margin: 0px;
            }

            @media screen and (min-width: 768px) {
                div,
                p {
                    &:nth-of-type(1) {
                        margin-right: 60px;
                    }
                }

                div p {
                    &:nth-of-type(1) {
                        margin-right: 10px;
                    }
                }

                .toggle-switch {
                    margin-right: 0px !important;
                }
            }
        }
    }

    &__settings {
        color: base.$primary-white;

        &__toggle-switch-wrapper,
        &__help-wrapper {
            display: flex;
            flex-direction: row;
            justify-content: space-between;
            align-items: center;

            p {
                margin-right: 10px;
            }
        }

        &__toggle-switch-wrapper {
            i {
                &:first-of-type {
                    margin-right: 7px;
                    font-size: 1.1rem;
                }
                &:last-of-type {
                    margin-left: 7px;
                    font-size: 1.2rem;
                }
            }
        }

        &__help-wrapper {
            cursor: pointer;

            i {
                font-size: 1.1rem;
            }
        }
    }

    &__info {
        color: base.$dark-grey;

        a {
            @include transition(0.3s);
            &:hover {
                color: base.$primary-white;
            }
        }
    }

    @media screen and (min-width: 1092px) {
    }
}
