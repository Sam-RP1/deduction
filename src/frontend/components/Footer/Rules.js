import React, { useMemo } from 'react';
// import PropTypes from 'prop-types';

import Chevron from '../UI/Indicators/Chevron/Chevron';

import './Rules.scss';

const Rules = () => {
    const toggleTab = (evt) => {
        const elem = evt.target.parentElement;
        const isCollapsed = elem.getAttribute('data-collapsed') === 'false';

        if (isCollapsed) {
            collapseTab(elem);
        } else {
            expandTab(elem);
        }
    };

    const collapseTab = (elem) => {
        const content = elem.childNodes[1];
        content.style.height = 0 + 'px';

        elem.setAttribute('data-collapsed', 'true');
    };

    const expandTab = (elem) => {
        const content = elem.childNodes[1];
        const contentHeight = content.scrollHeight;
        content.style.height = contentHeight + 10 + 'px';

        elem.setAttribute('data-collapsed', 'false');
    };

    return useMemo(() => {
        return (
            <div className='rules'>
                <div className='rules__tab' data-collapsed='true'>
                    <div id='deduction-info' className='rules__tab__title' onClick={(evt) => toggleTab(evt)}>
                        <h1>Deduction</h1>
                        <Chevron />
                    </div>
                    <div className='rules__tab__content'>
                        <p>
                            Deduction is a game where players are split into two teams, red and blue, and face off
                            against each other in an attempt to get their teams score down to 0 first!
                        </p>
                    </div>
                </div>
                <div className='rules__tab' data-collapsed='true'>
                    <div id='deduction-basics' className='rules__tab__title' onClick={(evt) => toggleTab(evt)}>
                        <h1>Basics</h1>
                        <Chevron />
                    </div>
                    <div className='rules__tab__content'>
                        <p>
                            Deduction requires at least 4 players to be played properly. Each team must have at least 1
                            player be an &apos;insider&apos; and the remaining players be &apos;agents&apos;. Each
                            randomly-generated standard gameboard has 25 words which break down to 9 red, 8 blue, 7
                            neutral and 1 orange (bomb). To win the game a team must get their score down to 0 before
                            the opposing team.
                        </p>
                    </div>
                </div>
                <div className='rules__tab' data-collapsed='true'>
                    <div id='deduction-roles' className='rules__tab__title' onClick={(evt) => toggleTab(evt)}>
                        <h1>Roles</h1>
                        <Chevron />
                    </div>
                    <div className='rules__tab__content'>
                        <p>
                            In Deduction, players can either be an &apos;insider&apos; or an &apos;agent&apos;. Please
                            note that each team must have at least 1 &apos;insider&apos; and at least 1
                            &apos;agent&apos;. The job of an &apos;insider&apos; is to each turn give a single
                            one-worded clue accompanied by a number to their team&apos;s agents. Nothing else. For
                            example &quot;food 3&quot;. The number indicates how many words that teams agents are
                            allowed to guess that turn and the one-worded clue is to help the agents figure out which
                            words on the board they need to guess. The job of an &apos;agent&apos; is to listen to the
                            information their teams &apos;insider&apos; gives them and then try to &apos;deduce&apos;
                            (figure out) which words they need to guess so that they can reduce their teams score. For
                            example, it is red teams turn. The red teams &apos;insider&apos; says &quot;food 3&quot;.
                            This means that red teams agents must make 3 guesses for that turn and try to guess words
                            related to food. &apos;Insiders&apos; cannot give &apos;agents&apos; one-worded clues that
                            contain part of or a full word to be gussed. For example &quot;stone 2&quot; would be
                            cheating if one of the words to be guessed was &quot;stone&quot; or contained
                            &quot;stone&quot;, like &quot;stonecutter&quot;.
                        </p>
                    </div>
                </div>
                <div className='rules__tab' data-collapsed='true'>
                    <div id='deduction-scoring' className='rules__tab__title' onClick={(evt) => toggleTab(evt)}>
                        <h1>Scoring</h1>
                        <Chevron />
                    </div>
                    <div className='rules__tab__content'>
                        <p>
                            If an &apos;agent&apos; correctly guesses a word, it will reduce their teams&apos; score by
                            1. If an &apos;agent&apos; incorrectly guesses a word that belongs to the opposing team,
                            their teams&apos; turn ends immediately and the opposing team&apos;s score will be reduced
                            by 1. If an &apos;agent&apos; incorrectly guesses a word that belongs to neither team, then
                            their teams&apos; turn ends immediately. If an &apos;agent&apos; is unfortunate enough to
                            incorrectly guess the bomb word, then the opposing team wins the game.
                        </p>
                    </div>
                </div>
            </div>
        );
    });
};

// Rules.propTypes = {};

export default Rules;
