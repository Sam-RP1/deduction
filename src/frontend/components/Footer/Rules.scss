@use '../../styles/base';
@import '../../styles/partials/color-themes';
@import '../../styles/mixins/transition';
@import '../../styles/mixins/transform';
@import '../../styles/mixins/userSelect';

.rules {
    position: relative;
    top: -500px;
    display: flex;
    flex-direction: column;
    height: auto;
    width: 100%;
    border-radius: 5px;
    box-shadow: base.$box--shadow--alt;

    @include themed() {
        color: t($text);
        background-color: t($bg-alt);
    }
    @include transition-theme(color 0.3s ease, background-color 0.3s ease, height 0.25s ease);

    &__tab {
        overflow: hidden;
        &:first-child {
            border-radius: 5px 5px 0px 0px;
        }
        &:last-child {
            border-radius: 0px 0px 5px 5px;
        }

        &__title {
            display: flex;
            align-items: center;
            justify-content: space-between;
            border-top: 1px solid;
            border-bottom: 1px solid;
            box-sizing: border-box;
            padding: 10px;
            cursor: pointer;

            h1 {
                margin: 0px;
                font-size: 2rem;
            }

            h1,
            .chevron {
                pointer-events: none;
                @include userSelect(none);
            }

            @include themed() {
                color: t($text);
                background-color: t($btn);
                border-color: t($bg);
            }
            @include transition-theme(color 0.3s ease, background-color 0.3s ease, border-color 0.3s ease);
        }

        &__content {
            display: flex;
            height: 0px;
            width: 100%;
            overflow: hidden;

            @include themed() {
                border-color: t($bg);
                background-color: t($bg-alt);
            }
            @include transition(height 0.3s ease, background-color 0.3s ease);

            p {
                font-size: 1.1rem;
                margin: 0px;
                box-sizing: border-box;
                padding: 10px;
            }
        }
    }
}

// X
// &:nth-child(1) {
//     @include transform(rotate(44deg) translate(-3px, 7px));
// }

// &:nth-child(2) {
//     @include transform(rotate(136deg) translate(4px, 6px));
// }
