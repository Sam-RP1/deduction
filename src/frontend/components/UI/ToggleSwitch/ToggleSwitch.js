import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import './ToggleSwitch.scss';

const ToggleSwitch = (props) => {
    const [isToggled, setToggleState] = useState(false);

    const toggle = () => {
        isToggled ? setToggleState(false) : setToggleState(true);
    };

    useEffect(() => {
        if (props.toggle) {
            setToggleState(true);
        }
    }, []);

    return (
        <div
            onClick={() => {
                props.function();
                toggle();
            }}
            className={'toggle-switch ' + (isToggled ? 'active' : '')}
        >
            <span></span>
        </div>
    );
};

ToggleSwitch.propTypes = {
    toggle: PropTypes.bool,
    function: PropTypes.func.isRequired,
};

ToggleSwitch.defaultProps = {
    toggle: false,
    function: () => console.log('[TOGGLE SWITCH] undefined toggle switch function'),
};

export default ToggleSwitch;
