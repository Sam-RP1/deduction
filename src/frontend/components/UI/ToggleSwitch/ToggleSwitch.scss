@use '../../../styles/base';
@import '../../../styles/mixins/transform';
@import '../../../styles/mixins/transition';

.toggle-switch {
    position: relative;
    display: inline-block;
    height: 28px;
    width: 65px;

    span {
        position: absolute;
        cursor: pointer;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
        box-sizing: border-box;
        border: 2px solid #fff;
        border-radius: 14px;
        background-color: #c7c7c7;
        @include transition(color 0.32s ease);
    }

    span:before {
        position: absolute;
        content: '';
        height: 22px;
        width: 22px;
        top: 1px;
        left: 1px;
        border-radius: 50%;
        background-color: white;
        @include transition(0.32s ease);
    }

    &.active {
        span {
            background-color: base.$selected;
        }

        span:before {
            @include transform(translateX(36px));
        }
    }
}
