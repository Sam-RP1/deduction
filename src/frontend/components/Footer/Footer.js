import React from 'react';
import { NavLink } from 'react-router-dom';
import PropTypes from 'prop-types';

import ToggleSwitch from '../UI/ToggleSwitch/ToggleSwitch';

import './Footer.scss';

const Footer = (props) => {
    return (
        <footer className='footer'>
            <div className='footer__settings footer__settings--row'>
                <div className='footer__settings__toggle-switch-wrapper'>
                    <p>Toggle Dark Mode</p>
                    <ToggleSwitch function={props.themeToggle} />
                </div>
                <NavLink to='/settings' onClick={() => console.log('[BUTTON CLICK] open settings menu')}>
                    <i className='fas fa-cog'></i>
                </NavLink>
            </div>

            <div className='footer__info footer__info--row'>
                <p>
                    Developed by{' '}
                    <a href='http://srenshawpanting.co.uk/'>
                        <u>SRP Designs</u>
                    </a>
                </p>
                <p>
                    <a href=''>
                        <u>Donate here</u>
                    </a>
                </p>
            </div>
        </footer>
    );
};

Footer.propTypes = {
    themeToggle: PropTypes.func,
};

export default Footer;
