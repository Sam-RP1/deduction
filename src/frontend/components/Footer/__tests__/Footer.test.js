import React from 'react';

// Enzyme Imports
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

// Component Imports
import Footer from '../Footer';
import { NavLink } from 'react-router-dom';
import ToggleSwitch from '../../UI/ToggleSwitch/ToggleSwitch';

// Enzyme configuration
configure({ adapter: new Adapter() });

// Test cases
describe('<Footer /> | Component', () => {
    it('should render <ToggleSwitch />', () => {
        const wrapper = shallow(<Footer />);
        expect(wrapper.find(<ToggleSwitch />));
    });

    it('should render <NavLink />', () => {
        const wrapper = shallow(<Footer />);
        expect(wrapper.find(<NavLink to='/settings' />));
    });

    it('should render <ToggleSwitch /> and <NavLink />', () => {
        const wrapper = shallow(<Footer />);
        expect(wrapper.find(<ToggleSwitch />));
        expect(wrapper.find(<NavLink to='/settings' />));
    });

    it('should render correctly', () => {
        const wrapper = shallow(<Footer />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
