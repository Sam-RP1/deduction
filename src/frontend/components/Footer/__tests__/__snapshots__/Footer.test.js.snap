// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer /> | Component should render correctly 1`] = `
<footer
  className="footer"
>
  <div
    className="footer__settings footer__settings--row"
  >
    <div
      className="footer__settings__toggle-switch-wrapper"
    >
      <p>
        Toggle Dark Mode
      </p>
      <ToggleSwitch
        function={[Function]}
      />
    </div>
    <i
      className="fas fa-cog"
      onClick={[Function]}
    />
  </div>
  <div
    className="footer__info footer__info--row"
  >
    <p>
      Developed by
       
      <a
        href="http://srenshawpanting.co.uk/"
      >
        <u>
          SRP Designs
        </u>
      </a>
    </p>
    <p>
      <a
        href=""
      >
        <u>
          Donate here
        </u>
      </a>
    </p>
  </div>
</footer>
`;
